src/plans/cache.jl

```diff
@@ -0,0 +1,172 @@
+#
+# A Cache for Objectives
+#
+@doc raw"""
+     SimpleCacheObjective{O<:AbstractManifoldGradientObjective{E,TC,TG}, P, T,C} <: AbstractManifoldGradientObjective{E,TC,TG}
+
+Provide a simple cache for an [`AbstractManifoldGradientObjective`](@ref) that is for a given point `p` this cache
+stores a point `p` and a gradient ``\operatorname{grad} f(p)`` in `X` as well as a cost value ``f(p)`` in `c`.
+
+Both `X` and `c` are accompanied by booleans to keep track of their validity.
+
+# Constructor
+
+    SimpleCacheObjective(M::AbstractManifold, obj::AbstractManifoldGradientObjective; kwargs...)
+
+## Keyword
+* `p` (`rand(M)`) – a point on the manifold to initialize the cache with
+* `X` (`get_gradient(M, obj, p)` or `zero_vector(M,p)`) – a tangent vector to store the gradient in, see also `initialize`
+* `c` (`get_cost(M, obj, p)` or `0.0`) – a value to store the cost function in `initialize`
+* `initialized` (`true`) – whether to initialize the cached `X` and `c` or not.
+"""
+mutable struct SimpleCacheObjective{
+    E<:AbstractEvaluationType,TC,TG,O<:AbstractManifoldGradientObjective{E,TC,TG},P,T,C
+} <: AbstractManifoldGradientObjective{E,TC,TG}
+    objective::O
+    p::P # a point
+    X::T # a vector
+    X_valid::Bool
+    c::C # a value for the cost
+    c_valid::Bool
+end
+
+function SimpleCacheObjective(
+    M::AbstractManifold,
+    obj::O;
+    initialized=true,
+    p=rand(M),
+    X=initialized ? get_gradient(M, obj, p) : zero_vector(M, p),
+    c=initialized ? get_cost(M, obj, p) : 0.0,
+) where {E<:AbstractEvaluationType,TC,TG,O<:AbstractManifoldGradientObjective{E,TC,TG}}
+    return SimpleCacheObjective{E,TC,TG,O,typeof(p),typeof(X),typeof(c)}(
+        obj, p, X, initialized, c, initialized
+    )
+end
+
+#
+# Default implementation
+#
+function get_cost(M::AbstractManifold, sco::SimpleCacheObjective, p)
+    if sco.p != p || !sco.c_valid
+        # else evaluate cost, invalidate grad if p changed
+        sco.c = get_cost(M, sco.objective, p)
+        # if we switched points, invalidate X
+        (sco.p != p) && (sco.X_valid = false)
+        sco.p = p
+        sco.c_valid = true
+    end
```

Comment:
```suggestion
    scop_neq_p = sco.p != p
    if scop_neq_p || !sco.c_valid
        # else evaluate cost, invalidate grad if p changed
        sco.c = get_cost(M, sco.objective, p)
        # if we switched points, invalidate X
        scop_neq_p && (sco.X_valid = false)
        # we need to copy in case the given point gets modified
        copyto!(M, sco.p, p)
        sco.c_valid = true
    end
```

Reply:
Is that comparison so expensive? If so we should change all other functions in that file as well.

Reply:
Oh and the copy breaks a lot of things, let's not do that. In the test it suddenly made p equal to q and that should not happen that p changes in the tests. So I am against the copyto, since it has strange outside side effects.

Reply:
Ah the side effects might come from me passing `p` in when constructing it, then we should at least do a copy in the constructor.

Reply:
It can be somewhat expensive, so I think all functions should be adapted like this.

> Oh and the copy breaks a lot of things, let's not do that. In the test it suddenly made p equal to q and that should not happen that p changes in the tests. So I am against the copyto, since it has strange outside side effects.

> Ah the side effects might come from me passing `p` in when constructing it, then we should at least do a copy in the constructor.

Yes, if the constructor makes a copy then there should be no problem with reusing that. If we don't do `copyto!` on a point, then imagine something like this:
```julia
p = some_point
c = get_cost(M, obj, p)
retract!(M, p, some_vector)
new_cost = get_cost(M, obj, p) # we get the old cached value `c` because `p` is still `p`
```

Reply:
good catch, but then we should really copy in the constructor, because otherwise the retract will also update the reference in the cache and you always get the initial (cached) cost.

Reply:
Yes, you're right.

Reply:
I adapted the code. We were both right to some extend that is nice :)

I kept your variable for the comparison, though it should not be necessary since the last comparison is done (on purpose) before the copyto! – together with the copy in the constructor this now works fine.